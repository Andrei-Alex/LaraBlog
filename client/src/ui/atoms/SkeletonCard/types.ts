export interface ISkeletonCard {
  width: string;
  height: string;
  vertical: boolean;
  testID: string;
  avatarWidth: string;
  avatarHeight: string;
}
