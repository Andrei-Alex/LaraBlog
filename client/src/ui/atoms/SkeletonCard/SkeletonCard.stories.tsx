import { Meta, StoryObj } from '@storybook/react';

import SkeletonCard from './SkeletonCard';

const meta = {
  title: 'Atoms/SkeletonCard',
  component: SkeletonCard,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    width: {
      name: 'Width',
      type: { name: 'string', required: false },
      control: 'text',
      description: 'Component width',
      table: {
        type: { summary: 'string' },
        defaultValue: { summary: '200px' },
      },
    },
    height: {
      name: 'Height',
      type: { name: 'string', required: false },
      control: 'text',
      description: 'Component height',
      table: {
        type: { summary: 'string' },
        defaultValue: { summary: '135px' },
      },
    },
    vertical: {
      name: 'Vertical',
      type: { name: 'string', required: false },
      control: 'boolean',
      description: 'Change the orientation of the card',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: 'false' },
      },
    },
    testID: {
      name: 'Test ID',
      type: { name: 'string', required: false },
      control: 'text',
      description: 'ID for testing library',
      table: {
        type: { summary: 'string' },
        defaultValue: { summary: 'skeletonCard' },
      },
    },
  },
  args: {
    width: '200px',
    height: '135px',
    vertical: false,
    testID: 'skeletonCard',
  },
} satisfies Meta<typeof SkeletonCard>;

export default meta;
export type Story = StoryObj<typeof meta>;
export const Primary: Story = {
  args: {
    width: '200px',
    height: '135px',
  },
};
