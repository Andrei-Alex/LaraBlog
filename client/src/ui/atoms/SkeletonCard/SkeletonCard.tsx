import React from 'react';

import { ISkeletonCard, styles } from '.';

/**
 * Renders a customizable skeleton card that can be used as a placeholder while content is loading.
 * This component is highly flexible, allowing for adjustments to its dimensions, orientation, and more
 * through a set of well-defined props. It's ideal for enhancing the user experience during data fetch operations
 * by providing a visual cue that content is on the way.
 *
 * The component leverages inline styling for dynamic customization and supports both horizontal and vertical layouts.
 * Additionally, a `testID` prop is provided to facilitate testing.
 *
 * ## Props
 *
 * - `width` (string): Specifies the overall width of the skeleton card. Defaults to "200px".
 * - `height` (string): Specifies the overall height of the skeleton card. Defaults to "135px".
 * - `avatarWidth` (string): Determines the width of the avatar placeholder within the card. Defaults to "48px".
 * - `avatarHeight` (string): Determines the height of the avatar placeholder within the card. Defaults to "48px".
 * - `vertical` (boolean): When true, adjusts the layout to a vertical orientation. Defaults to false for a horizontal layout.
 * - `testID` (string): Provides a unique identifier for the component, useful in testing environments. Defaults to "skeletonCard".
 *
 * ## Example Usage
 *
 * ```jsx
 * <SkeletonCard
 *   width="250px"
 *   height="150px"
 *   avatarWidth="50px"
 *   avatarHeight="50px"
 *   vertical={true}
 *   testID="mySkeletonCard"
 * />
 * ```
 *
 * @component
 *
 * @param {Partial<ISkeletonCard>} props - Destructured props object including optional width, height, avatarWidth, avatarHeight, vertical, and testID.
 *
 * @returns {JSX.Element} The `SkeletonCard` component rendered with applied styles and attributes based on props.
 */

function SkeletonCard({
  width = '200px',
  height = '135px',
  avatarWidth = '48px',
  avatarHeight = '48px',
  vertical = false,
  testID = 'skeletonCard',
}: Partial<ISkeletonCard>) {
  const style: React.CSSProperties = {
    width,
    height,
    flexDirection: vertical ? 'column' : 'row',
  };
  const AvatarStyle: React.CSSProperties = {
    marginBottom: vertical ? '12px' : '0px',
    minWidth: avatarWidth,
    minHeight: avatarHeight,
  };
  return (
    <div style={style} className={styles.card} data-testid={testID}>
      <div style={AvatarStyle} className={styles.avatar} />
      <div className={styles.content}>
        <div className={styles.line} />
        <div className={styles.line} />
        <div className={styles.line} />
      </div>
    </div>
  );
}
export default SkeletonCard;
