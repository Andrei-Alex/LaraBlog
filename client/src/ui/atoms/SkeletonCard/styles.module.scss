.card {
    @apply flex items-center justify-center mb-5 p-2 bg-gray-100 animate-pulse rounded;
    .avatar {
        @apply rounded-full bg-gray-300 mx-auto;
    }
    .content {
        width: 100%;
        @apply flex-grow space-y-3;
        .line {
            @apply mx-auto w-4/5 h-2.5 bg-gray-300 rounded;
        }
    }
}
