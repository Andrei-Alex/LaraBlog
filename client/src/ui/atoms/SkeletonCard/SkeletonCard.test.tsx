import React from 'react';
import { render, screen } from '@testing-library/react';
import SkeletonCard from './SkeletonCard';

describe('SkeletonCard', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<SkeletonCard />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders correctly', () => {
    render(<SkeletonCard />);
    const skeleton = screen.getByTestId('skeletonCard');
    expect(skeleton).toBeInTheDocument();
  });
  it('should apply given width from props', () => {
    const { container } = render(<SkeletonCard width="200px" />);
    expect(container.firstChild).toHaveStyle('width: 200px');
  });
  it('should apply given height from props', () => {
    const { container } = render(<SkeletonCard height="200px" />);
    expect(container.firstChild).toHaveStyle('height: 200px');
  });
  it('should have flex-direction column when vertical is true', () => {
    const { container } = render(<SkeletonCard vertical />);
    expect(container.firstChild).toHaveStyle('flex-direction: column');
  });

  it('should have flex-direction row when vertical is false', () => {
    const { container } = render(<SkeletonCard vertical={false} />);
    expect(container.firstChild).toHaveStyle('flex-direction: row');
  });
});
