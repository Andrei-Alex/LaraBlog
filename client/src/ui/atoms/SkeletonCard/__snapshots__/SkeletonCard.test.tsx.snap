// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkeletonCard renders correctly 1`] = `
<DocumentFragment>
  <div
    class="card"
    data-testid="skeletonCard"
    style="width: 200px; height: 135px; flex-direction: row;"
  >
    <div
      class="avatar"
      style="margin-bottom: 0px; min-width: 48px; min-height: 48px;"
    />
    <div
      class="content"
    >
      <div
        class="line"
      />
      <div
        class="line"
      />
      <div
        class="line"
      />
    </div>
  </div>
</DocumentFragment>
`;
